src/protocol/boolean/prefix_or.rs

```diff
@@ -0,0 +1,365 @@
+    ///   [y] = 0 1 1 1
+    ///   [f] = 0 1 0 0
+    ///   [b] = 0 0 0 0  0 0 1 1  1 1 1 1  1 1 1 1   // <- PrefixOr([a])
+    /// ```
+    fn step9(s: &[Replicated<B>], y: &[Replicated<B>], f: &[Replicated<B>]) -> Vec<Replicated<B>> {
+        let lambda = f.len();
+        (0..lambda)
+            .flat_map(|i| {
+                (0..lambda)
+                    .map(|j| s[lambda * i + j] + y[i] - f[i])
+                    .collect::<Vec<_>>()
+            })
+            .collect::<Vec<_>>()
+    }
+
+    /// Execute `PrefixOr`
+    #[allow(dead_code)]
+    #[allow(clippy::many_single_char_names)]
+    pub async fn execute(
+        &self,
+        ctx: ProtocolContext<'_, B>,
+    ) -> Result<Vec<Replicated<B>>, BoxError> {
+        // The paper assumes `l = λ^2`, where `l` is the bit length of the input
+        // share. Then the input is split into `λ` blocks each holding `λ` bits.
+        // Or operations are executed in parallel by running the blocks in
+        // parallel.
+        //
+        // The paper doesn't mention about cases where `l != λ^2`. For now, we
+        // pad the input with dummy bits at the end, and will be stripped off
+        // before returning the output to the caller. The output is correct
+        // regardless of the dummy bits, but may affect the performance if
+        // `λ^2 - l` becomes large. We should revisit this protocol once the
+        // prototype is complete, and optimize if necessary.
+        // TODO(taikiy): Prefix-Or optimization
+
+        // It should be safe to cast from `usize` to `u32` and back cast since
+        // the input we expect to operate on are 40-bit < `u8` << `u32`.
+        #[allow(clippy::cast_possible_truncation, clippy::cast_sign_loss)]
+        let lambda = f64::from(self.input.len() as u32).sqrt().ceil() as usize;
+        let dummy = vec![Replicated::new(B::ZERO, B::ZERO); lambda * lambda - self.input.len()];
+
+        let a = [self.input, &dummy].concat();
+        let x = Self::step1(&a, lambda, ctx.narrow(&Step::BlockOr1)).await?;
+        let y = Self::step2(&x, ctx.narrow(&Step::BlockOr2)).await?;
+        let f = Self::step3_4(&x, &y);
+        let g = Self::step5(&f, &a, ctx.narrow(&Step::Mult1)).await?;
+        let c = Self::step6(&g, lambda);
+        let b = Self::step7(&c, ctx.narrow(&Step::BlockOr3)).await?;
+        let s = Self::step8(&f, &b, ctx.narrow(&Step::Mult2)).await?;
+        let b = Self::step9(&s, &y, &f);
+
+        Ok(b[0..self.input.len()].to_vec())
+    }
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
+enum Step {
+    BlockOr1,
+    BlockOr2,
+    BlockOr3,
+    Mult1,
+    Mult2,
+}
+
+impl crate::protocol::Step for Step {}
+
+impl AsRef<str> for Step {
+    fn as_ref(&self) -> &str {
+        match self {
+            Self::BlockOr1 => "block_or_1",
+            Self::BlockOr2 => "block_or_2",
+            Self::BlockOr3 => "block_or_3",
+            Self::Mult1 => "mult_1",
+            Self::Mult2 => "mult_2",
```

Comment:
These names could be better =).

Reply:
OK, namings are hard... I changed them to (hopefully) more self-explanatory names.